homeassistant/components/ps4/config_flow.py

```diff
@@ -56,44 +61,74 @@ def __init__(self):
+                    errors[CONF_IP_ADDRESS] = 'no_ipaddress'
+            if not errors:
+                return await self.async_step_link()
+
+        mode_schema = OrderedDict()
+        mode_schema[vol.Required(
+            CONF_MODE, default=CONF_AUTO)] = vol.In(list(mode))
+        mode_schema[vol.Optional(CONF_IP_ADDRESS)] = str
+
+        return self.async_show_form(
+            step_id='mode',
+            data_schema=vol.Schema(mode_schema),
+            errors=errors,
+        )
+
     async def async_step_link(self, user_input=None):
         """Prompt user input. Create or edit entry."""
+        from pyps4_homeassistant.media_art import COUNTRIES
+        regions = sorted(COUNTRIES.keys())
         errors = {}
 
-        # Search for device.
-        devices = await self.hass.async_add_executor_job(
-            self.helper.has_devices)
+        if user_input is None:
+            # Search for device.
+            devices = await self.hass.async_add_executor_job(
+                self.helper.has_devices, self.m_device)
 
-        # Abort if can't find device.
-        if not devices:
-            return self.async_abort(reason='no_devices_found')
+            # Abort if can't find device.
+            if not devices:
+                return self.async_abort(reason='no_devices_found')
 
-        device_list = [
-            device['host-ip'] for device in devices]
+            self.device_list = [device['host-ip'] for device in devices]
+
+            # If entry exists check that devices found aren't configured.
+            if self.hass.config_entries.async_entries(DOMAIN):
+                for entry in self.hass.config_entries.async_entries(DOMAIN):
+                    conf_devices = entry.data['devices']
+                    for c_device in conf_devices:
+                        if c_device['host'] in self.device_list:
+                            # Remove configured device from search list.
+                            self.device_list.remove(c_device['host'])
 
-        # If entry exists check that devices found aren't configured.
-        if self.hass.config_entries.async_entries(DOMAIN):
-            for entry in self.hass.config_entries.async_entries(DOMAIN):
-                conf_devices = entry.data['devices']
-                for c_device in conf_devices:
-                    if c_device['host'] in device_list:
-                        # Remove configured device from search list.
-                        device_list.remove(c_device['host'])
             # If list is empty then all devices are configured.
-            if not device_list:
+            if not self.device_list:
                 return self.async_abort(reason='devices_configured')
 
         # Login to PS4 with user data.
         if user_input is not None:
             self.region = user_input[CONF_REGION]
             self.name = user_input[CONF_NAME]
-            self.pin = user_input[CONF_CODE]
+            self.pin = str(user_input[CONF_CODE])
```

Comment:
If we need it to be a string in the end, we can add `vol.Coerce(str)` to the end of the validation schema validators. We should try to have all input modifications in one place.

Reply:
Where do I put this? I tried adding it to the end of the vol.All tuple, but it overrides the int.